File: data-structures-and-algorithms/algorithm/leetCode/aboutArray/removeDuplicatesII/removeDuplicates.go
/**
 * @Author: shershon
 * @Description:
 * @Date: 2022/04/15 12:12 PM
 */

package removeDuplicates

// leetcode-80
// 删除排序数组中的重复项II
// 时间复杂度O(N)，空间复杂度O(1)
func removeDuplicates(nums []int) int {
	idx := 0
	for _, num := range nums {
		// 如果是前两位自然不可能重复两次以上（同时为了保证数组不越界）
		if (idx < 2) || (num != nums[idx-2]) {
			nums[idx] = num // 如果当前元素没有重复两次以上，我们就将其assign到目前不重复元素到达的index
			idx++
		}
	}
	return idx
}

File: data-structures-and-algorithms/algorithm/leetCode/aboutArray/removeDuplicates/removeDuplicates.php
<?php
/**
 * Created by PhpStorm.
 * User: Shershon
 * Date: 2019/8/29
 * Time: 14:48
 */

// leetcode-26
// 删除排序数组中的重复项
// 时间复杂度O(N^2)，空间复杂度O(1)
function removeDuplicates($arr)
{
    $idx = 0;
    while ($idx < count($arr) - 1) {
        if ($arr[$idx] == $arr[$idx + 1]) {
            unset($arr[$idx]);
        } else {
            $idx += 1;
        }
    }
    return count($arr);
}

// 优化，时间复杂度O(N)，空间复杂度O(1)
function removeDuplicates2($arr)
{
    $idx = 0;
    foreach ($arr as $v) {
        if ($idx < 1 || $v != $arr[$idx - 1]) {
            $arr[$idx] = $v;
            $idx += 1;
        }
    }
    return $idx;
}